networking: add download agent app

davidkostov opened the pull request:
Added a download agent application which downloads files given their url

mgenov commented on src/main/java/com/clouway/networking/downloadagent/DownloadAgent.java:
```diff
@@ -0,0 +1,50 @@
+        this.progressDisplayInterval = progressDisplayInterval;
+    }
+
+    public void downloadFile(String urlString, String savePath) {
```
You explicitly say that file name should be provided. It's better more common structures to be passed like InputStream and OutputStream as you are able to use memory for tests and real file streams in production.

mgenov commented on src/main/java/com/clouway/networking/downloadagent/DownloadAgent.java:
```diff
@@ -0,0 +1,50 @@
+            transferData = in.read();
+            while (transferData != -1) {
+                if (++counter >= progressDisplayInterval && counter % progressDisplayInterval == 0) {
+                    progressPercent = downloadProgress.getProgress();
```
When DownloadAgent transfers some bytes it is better to notify someone that something is transferred and he will be able to re-calculate the value of the progress.

mgenov commented on src/main/java/com/clouway/networking/downloadagent/DownloadAgent.java:
```diff
@@ -0,0 +1,50 @@
+            while (transferData != -1) {
+                if (++counter >= progressDisplayInterval && counter % progressDisplayInterval == 0) {
+                    progressPercent = downloadProgress.getProgress();
+                    System.out.println(progressPercent + "% downloaded.");
```
The system out should be removed as it's not role of the DownloadAgent to print the progress.

mgenov commented on src/main/java/com/clouway/networking/downloadagent/Main.java:
```diff
@@ -0,0 +1,10 @@
+ * @author David Kostov (david.kostov.cw@gmail.com)
+ */
+public class Main {
+    public static void main(String[] args) {
```
You have to wire everything and use it in the main of the application. 

You can use console for providing of the url for download.

mgenov commented on src/test/java/com/clouway/networking/DownloadAgentTests.java:
```diff
@@ -0,0 +1,51 @@
+                    returnValue(100)
+            ));
+        }});
+        downloadAgent.downloadFile("file:/home/clouway/workspaces/idea/Networking/src/test/test.jpg", "src/test/test2");
```
I won't be able to run this test as you are using real paths. 

ClassLoader and TemporaryFolder rule could be used for testing this. 

ClassLoaders allows resources to be loaded from the classpath, e.g you can add image in the same package as test (in the resources folder in case of maven/gradle). 

```
src/test/resources/com/clouway/downloadagent/source.jpg
src/test/java/com/clouway/downloadagent/DownloadAgentTest.java
```

davidkostov commented:
Fixed all issues, added DownloadProgress and RecordingProcessListener classes that implement the DownloadProgressListener interface. Moved the progress handling in the respective new classes leaving only the calculations within the DownloadAgent. Remade the testing for the application also added a main in which url can be given through console along with path to write the file to.

mgenov commented on src/main/java/com/clouway/networking/downloadagent/DownloadAgent.java:
```diff
@@ -0,0 +1,46 @@
+    private Double progressDisplayInterval;
+    private Integer progressPercent;
+
+    public DownloadAgent(Double progressDisplayInterval) {
```
This constructor need to be removed when no one uses it. The idea of the second constructor was to act as a bridge without compilation interruption to break your testing lifecycle.

mgenov commented on src/main/java/com/clouway/networking/downloadagent/DownloadAgent.java:
```diff
@@ -0,0 +1,46 @@
+
+    private final DownloadProgressListener progressListener;
+    private Double progressDisplayInterval;
+    private Integer progressPercent;
```
progressPercent should be moved as local variable

mgenov commented on src/test/java/com/clouway/networking/downloadagent/DownloadAgentTest.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Rule
+    public TemporaryFolder tempFolder = new TemporaryFolder();
+    private DownloadAgent downloadAgent = new DownloadAgent(/*progress,*/7500.0);
```
7500 doesn't give enough value. 

It would be better this part to be moved as constant in the DownloadAgent unless some friendly class name is used. PercentInterval.DOZENS, PercentInterval.ONE 
// 10'тки от процента
// на всеки 1 процент

davidkostov commented:
Fixed.

mgenov commented on src/main/java/com/clouway/networking/downloadagent/Main.java:
```diff
@@ -0,0 +1,36 @@
+    public static void main(String[] args) {
+        DownloadAgent dlAgent = new DownloadAgent(new DownloadProgress());
+        String urlString, savePath;
+        //left these paths for testing purposes
```
Commented code need to be removed.

davidkostov commented:
Fixed.
